src/GraphQL/Resolvers/ObservableFromAsyncEnumerable.cs

```diff
@@ -0,0 +1,140 @@
+using System.Runtime.CompilerServices;
+using System.Security.Claims;
+using GraphQL.Execution;
+using GraphQL.Instrumentation;
+using GraphQL.Types;
+using GraphQL.Validation;
+using GraphQLParser.AST;
+
+namespace GraphQL.Resolvers;
+
+/// <inheritdoc cref="ObservableFromAsyncEnumerable{T}.Create(Func{IResolveFieldContext, IAsyncEnumerable{T}})"/>
+internal sealed class ObservableFromAsyncEnumerable<T> : IObservable<object?>, IDisposable
+{
+    private readonly CancellationTokenSource _cancellationTokenSource;
+    private IAsyncEnumerable<T>? _enumerable;
+
+    private ObservableFromAsyncEnumerable(CancellationTokenSource cts, IAsyncEnumerable<T> enumerable)
+    {
+        _cancellationTokenSource = cts;
+        _enumerable = enumerable;
+    }
+
+    /// <summary>
+    /// Returns a source stream resolver delegate (which returns an <see cref="IObservable{T}"/>) from a delegate
+    /// which returns <see cref="IAsyncEnumerable{T}"/>. Each execution will create a new
+    /// <see cref="ObservableFromAsyncEnumerable{T}"/> instance which can only be used once.
+    /// </summary>
+    public static Func<IResolveFieldContext, ValueTask<IObservable<object?>>> Create(Func<IResolveFieldContext, ValueTask<IAsyncEnumerable<T>>> func)
+    {
+        return async context =>
+        {
+            var cts = CancellationTokenSource.CreateLinkedTokenSource(context.CancellationToken);
+            var enumerable = await func(new OverrideCancellationContext(context, cts.Token)).ConfigureAwait(false);
+            return new ObservableFromAsyncEnumerable<T>(cts, enumerable);
```

Comment:
The underlying cancellation token is signaled if the HTTP connection is aborted.  However, the cts here may be signalled if the specific subscription is stopped within a websocket connection.  Hence a linked token is created and passed into the `IResolveFieldContext` which is used to populate the arguments of the delegate.

Reply:
I prefer to place such words into code as comments.

Reply:
```suggestion
            var cts = CancellationTokenSource.CreateLinkedTokenSource(context.CancellationToken);

            // The underlying cancellation token is signaled if the HTTP connection is aborted. However, the
            // cts here may be signalled if the specific subscription is stopped within a websocket
            // connection. Hence a linked token is created and passed into the IResolveFieldContext
            // which is used to populate the arguments of the delegate.

            var enumerable = await func(new OverrideCancellationContext(context, cts.Token)).ConfigureAwait(false);
            return new ObservableFromAsyncEnumerable<T>(cts, enumerable);
```